### ISSUE TITLE
Always use CURIEs for identifier slots in schema

### ISSUE THREAD

**Issue by sujaypatil96** [microbiomedata/nmdc-schema#821]
A general best practice is to always use CURIEs for identifier slots on classes in the schema. Refer: https://berkeleybop.github.io/best_practice/

Some slots in the schema that are currently modelled differently from best practice:
* img_identifiers: https://microbiomedata.github.io/nmdc-schema/img_identifiers/
* alternative_identifiers: https://microbiomedata.github.io/nmdc-schema/alternative_identifiers/
* assembly_identifiers: https://microbiomedata.github.io/nmdc-schema/assembly_identifiers/
* gold_identifiers: https://microbiomedata.github.io/nmdc-schema/gold_identifiers/
* insdc_identifiers: https://microbiomedata.github.io/nmdc-schema/insdc_identifiers/
* insdc_bioproject_identifiers: https://microbiomedata.github.io/nmdc-schema/insdc_bioproject_identifiers/
* mgnify_project_identifiers: https://microbiomedata.github.io/nmdc-schema/mgnify_project_identifiers/
 
Since they are identifiers, plural, they all need to be modelled as:
```
range: uriorcurie
multivalued: true
```

**Comment by turbomam** [comment]
I agree, but we should be prepared for the impacts this will have on our legacy data, and on a our future data collection.

Will we need to provide validation patterns? At least to constrain the legal prefixes? I think we have already done this to some degree, esp for `insdc_identifiers`. (That goes beyond the non-enforced prefix suggestions).

**Comment by turbomam** [comment]
I'm trying to apply this to the `alternative_identifiers` parent property now in `src/schema/basic_slots.yaml`

@aclum had asked for something similar in 

- https://github.com/microbiomedata/nmdc-schema/issues/868#issuecomment-1563275100

**Comment by turbomam** [comment]
even in main, `make all` is failing now during `even in main, `make all` is failing in `check-jsonschema`

TL;DR

```
jsonschema.exceptions.UnknownType: Unknown type 'uriorcurie' for validator with schema:
    {'type': 'uriorcurie'}

While checking instance:
    'gold:Gs0110115'
```

longer version

```
Traceback (most recent call last):
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/jsonschema/validators.py", line 318, in is_type
    return self.TYPE_CHECKER.is_type(instance, type)
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/jsonschema/_types.py", line 119, in is_type
    raise UndefinedTypeCheck(type) from None
jsonschema.exceptions.UndefinedTypeCheck: Type 'uriorcurie' is unknown to this type checker

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/bin/check-jsonschema", line 8, in <module>
    sys.exit(main())
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/check_jsonschema/cli.py", line 285, in main
    execute(args)
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/check_jsonschema/cli.py", line 333, in execute
    ret = checker.run()
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/check_jsonschema/checker.py", line 87, in run
    self._run()
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/check_jsonschema/checker.py", line 77, in _run
    result = self._build_result()
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/check_jsonschema/checker.py", line 71, in _build_result
    for err in validator.iter_errors(data):
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/jsonschema/validators.py", line 288, in iter_errors
    for error in errors:
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/jsonschema/_validators.py", line 332, in properties
    yield from validator.descend(
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/jsonschema/validators.py", line 305, in descend
    for error in self.evolve(schema=schema).iter_errors(instance):
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/jsonschema/validators.py", line 288, in iter_errors
    for error in errors:
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/jsonschema/_legacy_validators.py", line 126, in items_draft6_draft7_draft201909
    yield from validator.descend(item, items, path=index)
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/jsonschema/validators.py", line 305, in descend
    for error in self.evolve(schema=schema).iter_errors(instance):
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/jsonschema/validators.py", line 288, in iter_errors
    for error in errors:
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/jsonschema/_validators.py", line 298, in ref
    yield from validator.descend(instance, resolved)
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/jsonschema/validators.py", line 305, in descend
    for error in self.evolve(schema=schema).iter_errors(instance):
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/jsonschema/validators.py", line 288, in iter_errors
    for error in errors:
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/jsonschema/_validators.py", line 332, in properties
    yield from validator.descend(
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/jsonschema/validators.py", line 305, in descend
    for error in self.evolve(schema=schema).iter_errors(instance):
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/jsonschema/validators.py", line 288, in iter_errors
    for error in errors:
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/jsonschema/_legacy_validators.py", line 126, in items_draft6_draft7_draft201909
    yield from validator.descend(item, items, path=index)
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/jsonschema/validators.py", line 305, in descend
    for error in self.evolve(schema=schema).iter_errors(instance):
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/jsonschema/validators.py", line 288, in iter_errors
    for error in errors:
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/jsonschema/_validators.py", line 321, in type
    if not any(validator.is_type(instance, type) for type in types):
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/jsonschema/_validators.py", line 321, in <genexpr>
    if not any(validator.is_type(instance, type) for type in types):
  File "/home/mark/.cache/pypoetry/virtualenvs/nmdc-schema-_1824IzR-py3.9/lib/python3.9/site-packages/jsonschema/validators.py", line 320, in is_type
    raise exceptions.UnknownType(type, instance, self.schema)
jsonschema.exceptions.UnknownType: Unknown type 'uriorcurie' for validator with schema:
    {'type': 'uriorcurie'}

While checking instance:
    'gold:Gs0110115'
make: *** [project.Makefile:320: jsonschema-check-all-valid-databases] Error 1
(nmdc-schema-py3.9) mark@NUC:~/gitrepos/nmdc-schema$ make all test
```

**Comment by aclum** [comment]
@turbomam  I think the issue is that the schema is currently enforcing caps for the curie prefix, your test uses lowercase. 

grep pattern external_identifiers.yaml | grep GOLD
    pattern: "^GOLD:Gs[0-9]+$"
    pattern: "^GOLD:Gb[0-9]+$"
    pattern: "^GOLD:Gp[0-9]+$"
    pattern: "^GOLD:Ga[0-9]+$"

As an aside this should be fixed at some point, identifiers.org allows for both 'GOLD and 'gold' see https://registry.identifiers.org/registry/gold

**Comment by turbomam** [comment]
THnaks, @aclum . This is realy helpful. I am a little nervous because those values with an uppercase GOLD prefix passed the validation tests form when I started until yesterday.  I'm going to take it slow in documenting this. I'm not sure what happened in the past 24 hours, unless I merged in something I wasn't aware of.

**Comment by aclum** [comment]
has this been addressed aside from legacy identifiers?

**Comment by turbomam** [comment]
yeah, I guess so. Thanks.
